signcrypt_seal.go

```diff
@@ -0,0 +1,262 @@
+	}
+
+	// Compute the identifier that the receiver will use to find this entry.
+	// Include the recipient index, so that this identifier is unique even if
+	// two recipients claim the same public key (though unfortunately that
+	// means that recipients will need to recompute the identifier for each
+	// entry in the recipients list). This identifier is somewhat redundant,
+	// because a recipient could instead just attempt to decrypt the payload
+	// key secretbox and see if it works, but including them adds a bit to
+	// anonymity by making box key recipients indistinguishable from symmetric
+	// key recipients.
+	keyIdentifierDigest := sha512.New()
+	keyIdentifierDigest.Write([]byte("saltpack signcryption derived key identifier\x00"))
+	keyIdentifierDigest.Write(derivedKey[:])
+	keyIdentifierDigest.Write(nonceForPayloadKeyBoxV2(index)[:])
+	identifier := keyIdentifierDigest.Sum(nil)[0:32]
+
+	payloadKeyBox := secretbox.Seal(
+		nil,
+		payloadKey[:],
+		(*[24]byte)(nonceForPayloadKeyBoxV2(index)),
+		(*[32]byte)(derivedKey))
+
+	return receiverKeys{
+		ReceiverKID:   identifier,
+		PayloadKeyBox: payloadKeyBox,
+	}
+}
+
+type ReceiverSymmetricKey struct {
+	// In practice these identifiers will be KBFS TLF keys.
+	Key SymmetricKey
+	// In practice these identifiers will be KBFS TLF pseudonyms.
+	Identifier []byte
+}
+
+func receiverEntryForSymmetricKey(receiverSymmetricKey ReceiverSymmetricKey, ephemeralPub BoxPublicKey, payloadKey SymmetricKey, index uint64) receiverKeys {
+	// Derive a message-specific shared secret by hashing the symmetric key and
+	// the ephemeral public key together. This lets us use nonces that are
+	// simple counters.
+	derivedKeyDigest := sha512.New()
+	derivedKeyDigest.Write([]byte(signcryptionSymmetricKeyContext))
+	derivedKeyDigest.Write(ephemeralPub.ToKID())
+	derivedKeyDigest.Write(receiverSymmetricKey.Key[:])
+	derivedKey, err := rawBoxKeyFromSlice(derivedKeyDigest.Sum(nil)[0:32])
+	if err != nil {
+		panic(err) // should be statically impossible, if the slice above is the right length
+	}
+
+	payloadKeyBox := secretbox.Seal(
+		nil,
+		payloadKey[:],
+		(*[24]byte)(nonceForPayloadKeyBoxV2(index)),
+		(*[32]byte)(derivedKey))
+
+	// Unlike the box key case, the identifier is supplied by the caller rather
+	// than computed. (These will be KBFS TLF pseudonyms.)
+	return receiverKeys{
+		ReceiverKID:   receiverSymmetricKey.Identifier,
+		PayloadKeyBox: payloadKeyBox,
+	}
+}
+
+func (sss *signcryptSealStream) init(receiverBoxKeys []BoxPublicKey, receiverSymmetricKeys []ReceiverSymmetricKey) error {
+	ephemeralKey, err := sss.keyring.CreateEphemeralKey()
+	if err != nil {
+		return err
+	}
+
+	eh := &SigncryptionHeader{
+		FormatName: SaltpackFormatName,
+		Version:    SaltpackVersion2,
+		Type:       MessageTypeSigncryption,
+		Ephemeral:  ephemeralKey.GetPublicKey().ToKID(),
+	}
+	sss.header = eh
+	if err := randomFill(sss.encryptionKey[:]); err != nil {
+		return err
+	}
+
+	// TODO: anonymous senders?
```

Comment:
We'll do an all zeroes key (and all zeroes signatures in the payload) to indicate anonymous mode.